packages/shell-api/src/bulk.ts

```diff
@@ -0,0 +1,217 @@
+    this._hint = hintDoc;
+    return this;
+  }
+
+  remove(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.remove();
+    return this._parentBulk;
+  }
+
+  removeOne(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.removeOne();
+    return this._parentBulk;
+  }
+
+  replaceOne(replacement: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(replacement);
+    const op = { ...replacement };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    this._innerFind.replaceOne(op);
+    return this._parentBulk;
+  }
+
+  updateOne(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.updateOne(op);
+    return this._parentBulk;
+  }
+
+  update(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.update(op);
+    return this._parentBulk;
+  }
+
+  upsert(): BulkFindOp {
+    assertArgsDefined();
+    this._innerFind.upsert();
+    return this;
+  }
+}
+
+
+@shellApiClassDefault
+@hasAsyncChild
+export default class Bulk extends ShellApiClass {
+  _mongo: Mongo;
+  _collection: any; // to avoid circular ref
+  _batchCounts: any;
+  _executed: boolean;
+  _batches: any;
+  _innerBulk: any;
+
+  constructor(collection, innerBulk) {
+    super();
+    this._collection = collection;
+    this._mongo = collection._mongo;
+    this._innerBulk = innerBulk;
+    this._batches = [];
+    this._batchCounts = {
+      nInsertOps: 0,
+      nUpdateOps: 0,
+      nRemoveOps: 0
+    };
+  }
+
+  /**
+   * Internal method to determine what is printed for this class.
+   */
+  _asPrintable(): string {
+    return this.tojson();
+  }
+
+  /**
+   * Internal helper for emitting collection API call events.
+   *
+   * @param methodName
+   * @param methodArguments
+   * @private
+   */
+  private _emitBulkApiCall(methodName: string, methodArguments: Document = {}): void {
+    this._mongo._internalState.emitApiCall({
+      method: methodName,
+      class: 'Bulk',
+      db: this._collection._database._name,
+      coll: this._collection._name,
+      arguments: methodArguments
+    });
+  }
+
+  @returnsPromise
+  async execute(writeConcern?: WriteConcern): Promise<BulkWriteResult> {
+    if (this._executed) {
+      throw new MongoshInvalidInputError('A bulk operation cannot be re-executed');
+    }
+
+    if (this._innerBulk.s !== undefined && Array.isArray(this._innerBulk.s.batches)) {
+      this._batches = [...this._innerBulk.s.batches];
```

Comment:
do we need to keep track of the batches in a class property or we should rather have a `_getAllBatches` method so we always use the service provider state as source of truth?

Reply:
You can only call getOperations after a bulk op is executed, but the bulk op class for the driver deletes the batches attribute after it's executed. So we need to save the batches locally right before we execute. Just a difference from driver to shell api.